_source/_posts/2020-07-02-spring-boot-using-java-modules.md

```diff
@@ -0,0 +1,473 @@
+
+**Java modules** is another name for JPMS (Java Platform Module System), added on JDK 9 under the name Project Jigsaw. It allows applications (packaged as JAR or WAR) to define a `module-info.java` file on its root with directives that control which classes the application will allow others to access and which other modules it needs on compile or runtime.
+
+## Writing Initial Application Without Java Modules
+
+Let's write some classes without using the Java modules so we can afterward include the module definitions and see the differences. 
+What defines that an application is using Java modules is the presence of `module-info.java` on its source root. We'll be creating this later on.
+
+### Persistence Module
+
+Create a class `Bird` on the Persistence module for representing the entity that we'll be saving to DB. 
+This class will be stored on `persistence/src/main/java/com/okta/developer/animals/bird/Bird.java`
+
+```java
+package com.okta.developer.animals.bird;
+
+import org.springframework.data.annotation.Id;
+
+public class Bird {
+
+    @Id
+    private String id;
+
+    private String specie;
+    private String size;
+
+    public String getId() {
+        return id;
+    }
+
+    public void setId(String id) {
+        this.id = id;
+    }
+
+    public String getSpecie() {
+        return specie;
+    }
+
+    public void setSpecie(String specie) {
+        this.specie = specie;
+    }
+
+    public String getSize() {
+        return size;
+    }
+
+    public void setSize(String size) {
+        this.size = size;
+    }
+}
+
+```
+
+Now we need to create a repository to save this entity to DB. `Spring Data MongoDB` does that for us automatically creating the CRUD operations so we just have to create an interface extending MongoRepository.
+This class will be stored on `persistence/src/main/java/com/okta/developer/animals/bird/BirdRepository.java`
+```java
+package com.okta.developer.animals.bird;
+
+import org.springframework.data.mongodb.repository.MongoRepository;
+
+public interface BirdRepository extends MongoRepository<Bird, String> {
+}
+```
+
+At last, for the persistence module, we'll be creating a service class to expose the persistence operations.
+This class will be stored on `persistence/src/main/java/com/okta/developer/animals/bird/BirdPersistence.java`
+
+```java
+package com.okta.developer.animals.bird;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import javax.annotation.PostConstruct;
+import java.util.List;
+
+@Component
+public class BirdPersistence {
+
+    private BirdRepository birdRepository;
+
+    @Autowired
+    public BirdPersistence(BirdRepository birdRepository) {
+        this.birdRepository = birdRepository;
+    }
+
+    @PostConstruct
+    void postConstruct(){
+        Bird sampleBird = new Bird();
+        sampleBird.setSpecie("Hummingbird");
+        sampleBird.setSize("small");
+        save(sampleBird);
```

Comment:
Great example, but the changes are persisted and rewritten when the app restarts
consider deleting entries before adding a new record 🤔 

```json
[{"id":"5f0c806e02bcd44eeed9af1c","specie":"Hummingbird","size":"small"},{"id":"5f0c81765c1e40081542ece5","specie":"Hummingbird","size":"small"},{"id":"5f0c81d67df83b42c643f521","specie":"Hummingbird","size":"small"},{"id":"5f0c827629c7cb426382ed9e","specie":"Hummingbird","size":"small"}]
```